kbkdf/src/lib.rs

```diff
@@ -0,0 +1,466 @@
+                    &(<KbkdfCore<K::KeySize, Prf::OutputSize> as KbkdfUser>::L::U32).to_be_bytes()
+                        [..],
+                );
+            }
+
+            let buf = h.finalize().into_bytes();
+            ki = Some(buf.clone());
+
+            let remaining = usize::min(buf.len(), builder.len());
+
+            builder[..remaining].copy_from_slice(&buf[..remaining]);
+            builder = &mut builder[remaining..];
+        }
+
+        assert_eq!(builder.len(), 0, "output has uninitialized bytes");
+
+        Ok(output)
+    }
+
+    /// Input the IV in the PRF
+    fn input_iv(&self, _ki: &mut Option<Array<u8, Prf::OutputSize>>) {}
+
+    /// Whether the KI should be reinjected every round.
+    const FEEDBACK_KI: bool = false;
+
+    const DOUBLE_PIPELINE: bool = false;
+}
+
+pub struct Counter<Prf, K, R = U32> {
+    _marker: PhantomData<(Prf, K, R)>,
+}
+
+impl<Prf, K, R> Default for Counter<Prf, K, R> {
+    fn default() -> Self {
+        Self {
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<Prf, K, R> Kbkdf<Prf, K, R> for Counter<Prf, K, R>
+where
+    Prf: Mac + KeyInit,
+    K: KeySizeUser,
+    K::KeySize: ArraySize + Mul<U8>,
+    <K::KeySize as Mul<U8>>::Output: Unsigned,
+    Prf::OutputSize: ArraySize + Mul<U8>,
+    <Prf::OutputSize as Mul<U8>>::Output: Unsigned,
+    R: sealed::R,
+{
+}
+
+pub struct Feedback<'a, Prf, K, R = U32>
+where
+    Prf: Mac,
+{
+    iv: Option<&'a Array<u8, Prf::OutputSize>>,
+    _marker: PhantomData<(Prf, K, R)>,
+}
+
+impl<'a, Prf, K, R> Feedback<'a, Prf, K, R>
+where
+    Prf: Mac,
+{
+    pub fn new(iv: Option<&'a Array<u8, Prf::OutputSize>>) -> Self {
+        Self {
+            iv,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<'a, Prf, K, R> Kbkdf<Prf, K, R> for Feedback<'a, Prf, K, R>
+where
+    Prf: Mac + KeyInit,
+    K: KeySizeUser,
+    K::KeySize: ArraySize + Mul<U8>,
+    <K::KeySize as Mul<U8>>::Output: Unsigned,
+    Prf::OutputSize: ArraySize + Mul<U8>,
+    <Prf::OutputSize as Mul<U8>>::Output: Unsigned,
+    R: sealed::R,
+{
+    fn input_iv(&self, ki: &mut Option<Array<u8, Prf::OutputSize>>) {
+        if let Some(iv) = self.iv {
+            *ki = Some(iv.clone())
+        }
+    }
+
+    const FEEDBACK_KI: bool = true;
+}
+
+pub struct DoublePipeline<Prf, K, R = U32>
+where
+    Prf: Mac,
+{
+    _marker: PhantomData<(Prf, K, R)>,
+}
+
+impl<Prf, K, R> Default for DoublePipeline<Prf, K, R>
+where
+    Prf: Mac,
+{
+    fn default() -> Self {
+        Self {
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<Prf, K, R> Kbkdf<Prf, K, R> for DoublePipeline<Prf, K, R>
+where
+    Prf: Mac + KeyInit,
+    K: KeySizeUser,
+    K::KeySize: ArraySize + Mul<U8>,
+    <K::KeySize as Mul<U8>>::Output: Unsigned,
+    Prf::OutputSize: ArraySize + Mul<U8>,
+    <Prf::OutputSize as Mul<U8>>::Output: Unsigned,
+    R: sealed::R,
+{
+    const DOUBLE_PIPELINE: bool = true;
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
These tests are relatively big, so it may be worth to move them into a separate file, e.g. `counter_tests.rs`.

Reply:
https://github.com/TheBestTvarynka/KDFs/pull/6

Reply:
done